From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 16 Sep 2021 22:00:39 +0200
Subject: [PATCH] Add per-player chunk send rate


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3e8cd3971ac8256a40d9b85cd7514998c965512c..575e37a9565edd0f8f7b13d5a7630cff594d9099 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1940,6 +1940,28 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param viewDistance view distance in [2, 32] or -1
      */
     public void setSendViewDistance(int viewDistance);
+
+    // Martijn in Paper start - per-player chunk send rate
+    /**
+     * Gets the target chunk send rate (in chunks per second) for this player.
+     * <p>
+     * The maximum chunk send rate an individual player will ever have. A value of -1 means the limit is disabled.
+     * </p>
+     * @return The target chunk send rate for this player, always either -1 (to indicate an unlimited chunk send rate) or a positive value >= 1
+     */
+    public double getTargetChunkSendRate();
+
+    /**
+     * Sets the target chunk send rate (in chunks per second) for this player.
+     * <p>
+     * The maximum chunk send rate an individual player will ever have. A value of -1 means the limit is disabled.
+     * <br>
+     * Any negative value given will be treated as -1 (disabling the limit). Any value in the range [0, 1] will be treated as 1. If given null, this method acts as unsetting the value for this player: instead, the value in the Paper configuration will be used.
+     * </p>
+     * @param chunkSendRate The target chunk send rate for this player, or null to defer to the value in the Paper configuration.
+     */
+    public void setTargetChunkSendRate(@Nullable Double chunkSendRate);
+    // Martijn in Paper end - per-player chunk send rate
     // Paper end
 
     /**
