From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 16 Sep 2021 15:39:49 +0200
Subject: [PATCH] Add entity tracking view distance


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index dc497606ebf9903c2d32276ecddda3231457346a..ab268f83b8f557b5d319aa78aa1927e18ac6381d 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -3603,6 +3603,44 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      * @param viewDistance view distance in [2, 32] or -1
      */
     public void setSendViewDistance(int viewDistance);
+
+    // Martijn in Paper start - entity tracking view distance
+    /**
+     * Gets the entity tracking view distance for this world.
+     * <p>
+     * Entity tracking view distance is the view distance where entities will be tracked (visible) for players in this world.
+     * </p>
+     * @return The entity tracking view distance for this world.
+     */
+    public int getEntityTrackingViewDistance();
+
+    /**
+     * Sets the entity tracking view distance for this world.
+     * <p>
+     * Entity tracking view distance is the view distance where entities will be tracked (visible) for players in this world.
+     * </p>
+     * @param viewDistance view distance >= 2, or -1
+     */
+    public void setEntityTrackingViewDistance(int viewDistance);
+
+    /**
+     * Gets whether to allow entity tracking beyond the no-tick view distance for this world.
+     * <p>
+     * Allowing entity tracking beyond the no-tick view distance can be used to allow players in this world to see entities in chunks that are not knowingly sent to the client by the server implementation, but by plugins in a custom manner.
+     * </p>
+     * @return Whether to allow entity tracking beyond the no-tick view distance for this world.
+     */
+    public boolean getAllowEntityTrackingBeyondNoTickViewDistance();
+
+    /**
+     * Sets whether to allow entity tracking beyond the no-tick view distance for this world.
+     * <p>
+     * Allowing entity tracking beyond the no-tick view distance can be used to allow players in this world to see entities in chunks that are not knowingly sent to the client by the server implementation, but by plugins in a custom manner.
+     * </p>
+     * @param allow Whether to allow entity tracking beyond the no-tick view distance for this world
+     */
+    public void setAllowEntityTrackingBeyondNoTickViewDistance(boolean allow);
+    // Martijn in Paper end - entity tracking view distance
     // Paper end - view distance api
     // Spigot start
     public class Spigot {
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3e8cd3971ac8256a40d9b85cd7514998c965512c..91f0880ce4a8a3b864736410815ef5827aaa99e4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1940,6 +1940,44 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param viewDistance view distance in [2, 32] or -1
      */
     public void setSendViewDistance(int viewDistance);
+
+    // Martijn in Paper start - entity tracking view distance
+    /**
+     * Gets the entity tracking view distance for this player.
+     * <p>
+     * Entity tracking view distance is the view distance where entities will be tracked (visible) for players.
+     * </p>
+     * @return The entity tracking view distance for this player.
+     */
+    public int getEntityTrackingViewDistance();
+
+    /**
+     * Sets the entity tracking view distance for this player.
+     * <p>
+     * Entity tracking view distance is the view distance where entities will be tracked (visible) for players.
+     * </p>
+     * @param viewDistance view distance >= 2, or -1
+     */
+    public void setEntityTrackingViewDistance(int viewDistance);
+
+    /**
+     * Gets whether to allow entity tracking beyond the no-tick view distance for this player.
+     * <p>
+     * Allowing entity tracking beyond the no-tick view distance can be used to allow players to see entities in chunks that are not knowingly sent to the client by the server implementation, but by plugins in a custom manner.
+     * </p>
+     * @return Whether to allow entity tracking beyond the no-tick view distance for this player.
+     */
+    public boolean getAllowEntityTrackingBeyondNoTickViewDistance();
+
+    /**
+     * Sets whether to allow entity tracking beyond the no-tick view distance for this player.
+     * <p>
+     * Allowing entity tracking beyond the no-tick view distance can be used to allow players to see entities in chunks that are not knowingly sent to the client by the server implementation, but by plugins in a custom manner.
+     * </p>
+     * @param allow A boolean indicating whether to allow entity tracking beyond the no-tick view distance for this player, or anull to unset a specific allowance for this player ({@link org.bukkit.World#getAllowEntityTrackingBeyondNoTickViewDistance will be used instead}
+     */
+    public void setAllowEntityTrackingBeyondNoTickViewDistance(@Nullable Boolean allow);
+    // Martijn in Paper end - entity tracking view distance
     // Paper end
 
     /**
