From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Sep 2021 15:39:06 +0200
Subject: [PATCH] Support lower chunk send rates


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 27cc92e989a55bfc72b3743f5479731f9067bf62..524ecc9d1afbbbfb5c208d2a1f32adc3ab3d4726 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1983,7 +1983,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * <p>
      * The maximum chunk send rate an individual player will ever have. A value of -1 means the limit is disabled.
      * </p>
-     * @return The target chunk send rate for this player, always either -1 (to indicate an unlimited chunk send rate) or a positive value >= 1
+     * @return The target chunk send rate for this player, always either -1 (to indicate an unlimited chunk send rate) or a positive value &gt;= 10<sup>-6</sup>
      */
     public double getTargetChunkSendRate();
 
@@ -1992,7 +1992,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * <p>
      * The maximum chunk send rate an individual player will ever have. A value of -1 means the limit is disabled.
      * <br>
-     * Any negative value given will be treated as -1 (disabling the limit). Any value in the range [0, 1] will be treated as 1. If given null, this method acts as unsetting the value for this player: instead, the value in the Paper configuration will be used.
+     * Any negative value given will be treated as -1 (disabling the limit). Any value in the range [0, 10<sup>-6</sup>] will be treated as 10<sup>-6</sup>. If given null, this method acts as unsetting the value for this player: instead, the value in the Paper configuration will be used.
      * </p>
      * @param chunkSendRate The target chunk send rate for this player, or null to defer to the value in the Paper configuration.
      */
