From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Fri, 10 Jun 2022 00:26:47 +0200
Subject: [PATCH] Fake server version


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index fc16e897f4ebaf9d79e49774e15a1cded33ef725..9b658b63cdcb6a2ad556b22ed26965a32cd51242 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -120,7 +120,14 @@ public final class Bukkit {
         if ("master".equals(gitBranch) || "main".equals(gitBranch)) {
             branchMsg = "";  // Don't show branch on main/master
         }
-        return "This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitCommit + branchMsg + ")";
+        // Martijn start - configurable server version
+        String gitCommitAndBranchMsg = gitCommit + branchMsg;
+        String fakeGitCommitAndBranchMsg = Bukkit.getServer().getFakeVersionGitCommit();
+        if (fakeGitCommitAndBranchMsg != null) {
+            gitCommitAndBranchMsg = fakeGitCommitAndBranchMsg;
+        }
+        return "This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitCommitAndBranchMsg + ")";
+        // Martijn end - configurable server version
         // Paper end
     }
 
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 30b49aed62fa67276e8964922ea3f84458d854bb..04ef354854573c9549706a205e00a352a309b562 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -104,6 +104,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public String getVersion();
 
+    // Martijn start - configurable server version
+    /**
+     * Gets a configured fake version Git commit and branch.
+     *
+     * @return the configured fake text, or null if no replacement was configured
+     */
+    @Nullable
+    public String getFakeVersionGitCommit();
+    // Martijn end - configurable server version
+
     /**
      * Gets the Bukkit version that this server is running.
      *
