From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sat, 2 Jul 2022 00:10:15 +0200
Subject: [PATCH] Fix QUICK_CRAFT clicks disappearing items


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 52aa304b2d7e6680a0cb6ffa9b7c4495afb9a781..51803531169b4875df758d732be57e5a77bd8399 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4107,7 +4107,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
         this.legacy = this.name().startsWith(LEGACY_PREFIX);
         // Martijn start - custom blocks and items
         String lowerCaseName = this.name().toLowerCase(Locale.ROOT);
-        if (lowerCaseName.startsWith("SUCRAFT_")) {
+        if (this.name().startsWith("SUCRAFT_")) {
             this.key = NamespacedKey.sucraft(lowerCaseName.substring(8));
         } else {
             this.key = NamespacedKey.minecraft(lowerCaseName);
