From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Tue, 19 Jul 2022 22:43:49 +0200
Subject: [PATCH] Z


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 781f1305cbfb5b84377b2a84a4cbafbd572cbc62..d692ae2d2eb5db188d77caa21c74ce3437ddeebd 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4329,7 +4329,8 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
     SUCRAFT_PAPER_BLOCK(78399, WHITE_WOOL),
     SUCRAFT_PAPER_SLAB(78400, SUCRAFT_WHITE_WOOL_SLAB),
     SUCRAFT_QUARTZ_SPIRAL_MEANDER(78401, CHISELED_QUARTZ_BLOCK),
-    SUCRAFT_QUARTZ_SPIRAL_MEANDER_SLAB(78402, SUCRAFT_CHISELED_QUARTZ_SLAB);
+    SUCRAFT_QUARTZ_SPIRAL_MEANDER_SLAB(78402, SUCRAFT_CHISELED_QUARTZ_SLAB),
+    SUCRAFT_ENDER_PEARL_BLOCK(78403, AMETHYST_BLOCK);
 
     // Martijn end - custom blocks and items
     //</editor-fold>
