From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 3 Jul 2022 00:33:42 +0200
Subject: [PATCH] Add Dirty Stone Bricks


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 51803531169b4875df758d732be57e5a77bd8399..5482a19178cc61095a18c9bd2c8e5120b1a6f99a 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4057,7 +4057,8 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
     LEGACY_RECORD_12(2267, 1),
 
     // Martijn start - custom blocks and items
-    SUCRAFT_SNOWY_STONE_BRICKS(78131, MOSSY_STONE_BRICKS);
+    SUCRAFT_SNOWY_STONE_BRICKS(78131, MOSSY_STONE_BRICKS),
+    SUCRAFT_DIRTY_STONE_BRICKS(78132, SUCRAFT_SNOWY_STONE_BRICKS);
     // Martijn end - custom blocks and items
     ;
     //</editor-fold>
