From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 3 Jul 2022 20:32:50 +0200
Subject: [PATCH] Add infested custom stone bricks


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 5482a19178cc61095a18c9bd2c8e5120b1a6f99a..5122d5c9c61b51fa206ff6f6768a2670c930df52 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4058,7 +4058,9 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
 
     // Martijn start - custom blocks and items
     SUCRAFT_SNOWY_STONE_BRICKS(78131, MOSSY_STONE_BRICKS),
-    SUCRAFT_DIRTY_STONE_BRICKS(78132, SUCRAFT_SNOWY_STONE_BRICKS);
+    SUCRAFT_DIRTY_STONE_BRICKS(78132, SUCRAFT_SNOWY_STONE_BRICKS),
+    SUCRAFT_INFESTED_SNOWY_STONE_BRICKS(78133, INFESTED_MOSSY_STONE_BRICKS),
+    SUCRAFT_INFESTED_DIRTY_STONE_BRICKS(78134, SUCRAFT_INFESTED_SNOWY_STONE_BRICKS);
     // Martijn end - custom blocks and items
     ;
     //</editor-fold>
@@ -4129,7 +4131,7 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
     }
 
     // Martijn start - custom blocks and items
-    private Material(final int id, Material vanillaTarget) {
+    Material(final int id, Material vanillaTarget) {
         this(id);
         this.vanillaTarget = vanillaTarget;
     }
