From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 7 Jul 2022 16:51:09 +0200
Subject: [PATCH] Create base code for multi-state custom blocks


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8109ff8c611d8539210378a0e5412279bd8b572e..bead606baa4f28a2c1e40c945977c996d13b7872 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2499,6 +2499,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     public boolean hasSuCraftResourcePack();
 
+    public boolean hasHighPingForCustomBlocks();
+
     // Martijn end - custom blocks and items
 
     /**
