From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Tue, 12 Jul 2022 12:30:52 +0200
Subject: [PATCH] In progress adding fake slabs


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 16c979a584ce8101d76c96f69c60b0355b834c73..becd918ce8f20d293af90a2e9e25ac18337b60a0 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4125,7 +4125,8 @@ public enum Material implements Keyed, net.kyori.adventure.translation.Translata
     SUCRAFT_RED_MAPLE_LEAVES(78195, SUCRAFT_RED_OAK_LEAVES),
     SUCRAFT_ORANGE_MAPLE_LEAVES(78196, SUCRAFT_RED_MAPLE_LEAVES),
     SUCRAFT_YELLOW_MAPLE_LEAVES(78197, SUCRAFT_RED_MAPLE_LEAVES),
-    SUCRAFT_DIRT_SLAB(78198, MUD_BRICK_SLAB);
+    SUCRAFT_DIRT_SLAB(78198, MUD_BRICK_SLAB),
+    SUCRAFT_SNOWY_STONE_BRICK_SLAB(78199, STONE_BRICK_SLAB);
 
     // Martijn end - custom blocks and items
     //</editor-fold>
