From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Sep 2021 15:38:51 +0200
Subject: [PATCH] Support lower chunk send rates


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index efa87b87b97b980bfea7a05a258c7c598acb2f3e..ec9a625b44b2e0368af02c8f274736326f5d3eaf 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -307,13 +307,15 @@ public final class PlayerChunkLoader {
     // Martijn in Paper start - per-player chunk send rate
     protected long getTargetSendPerPlayerAddend(ServerPlayer player) {
         final double chunkSendRate = player.getBukkitEntity().getTargetChunkSendRate();
-        // disable the limit for any given negative rate (with -1 the intended value to denote this)
+        // disable the limit for any given negative rate (with -1 the intended value to denote this, and 0L the appropriate output of this method in that case)
         return chunkSendRate < 0 ? 0L : (long) Math.round(1.0e9 / chunkSendRate);
     }
     // Martijn in Paper end - per-player chunk send rate
 
     protected long getMaxSendAddend() {
-        return PaperConfig.globalMaxChunkSendRate <= 1.0 ? 0L : (long)Math.round(1.0e9 / PaperConfig.globalMaxChunkSendRate);
+        // Martijn in Paper start - allow low chunk send rates
+        return PaperConfig.globalMaxChunkSendRate < 0 ? 0L : (long)Math.round(1.0e9 / PaperConfig.globalMaxChunkSendRate);
+        // Martijn in Paper end - allow low chunk send rates
     }
 
     public void onChunkPlayerTickReady(final int chunkX, final int chunkZ) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 64c9c67a3691562f6b8eefb43f9c4be35dbb1fa8..f156e208cc040b1a46ff762670da45a701124bb2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -588,7 +588,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         final Double playerTargetChunkSendRate = this.getHandle().targetChunkSendRate;
         final double effectiveTargetChunkSendRate = playerTargetChunkSendRate == null ? PaperConfig.playerTargetChunkSendRate : playerTargetChunkSendRate;
         // Make sure the returned rate is -1, or >= 1
-        return effectiveTargetChunkSendRate < 0 ? -1 : (effectiveTargetChunkSendRate <= 1 ? 1 : effectiveTargetChunkSendRate);
+        return effectiveTargetChunkSendRate < 0 ? -1 : Math.max(effectiveTargetChunkSendRate, 1.0e-6);
     }
 
     @Override
