From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Sep 2021 23:38:25 +0200
Subject: [PATCH] Update chunk send deadline based on chunk send rate


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index f07920ab954cd8eb4e06776322c92e178de8500d..885ab3870d5f2d11ed8641a8ecc0c512a8840fd7 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -305,7 +305,7 @@ public final class PlayerChunkLoader {
     }
 
     // Martijn in Paper start - per-player chunk send rate
-    protected long getTargetSendPerPlayerAddend(ServerPlayer player) {
+    public long getTargetSendPerPlayerAddend(ServerPlayer player) { // Martijn - update chunk send deadline based on chunk send rate
         final double chunkSendRate = player.getBukkitEntity().getTargetChunkSendRate();
         // disable the limit for any given negative rate (with -1 the intended value to denote this, and 0L the appropriate output of this method in that case)
         return chunkSendRate < 0 ? 0L : (long) Math.round(1.0e9 / chunkSendRate);
@@ -680,7 +680,7 @@ public final class PlayerChunkLoader {
         protected boolean usingLookingPriority;
 
         protected final ServerPlayer player;
-        protected final PlayerChunkLoader loader;
+        public final PlayerChunkLoader loader; // Martijn in Paper - update chunk send deadline based on chunk send rate
 
         // warning: modifications of this field must be aware that the loadQueue inside PlayerChunkLoader uses this field
         // in a comparator!
@@ -705,7 +705,7 @@ public final class PlayerChunkLoader {
         protected int loadViewDistance = -1;
         protected int tickViewDistance = -1;
 
-        protected long nextChunkSendTarget;
+        public long nextChunkSendTarget; // Martijn in Paper - update chunk send deadline based on chunk send rate
 
         public PlayerLoaderData(final ServerPlayer player, final PlayerChunkLoader loader) {
             this.player = player;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 9130dc9169d95c744d77e12d746b58876e2670fe..393601902802d947ccf42d0e94f1678f1446f24c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -594,6 +594,19 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public void setTargetChunkSendRate(Double chunkSendRate) {
         this.getHandle().targetChunkSendRate = chunkSendRate;
+        // Martijn start - update chunk send deadline based on chunk send rate
+        net.minecraft.server.level.ChunkMap chunkMap = this.getHandle().getLevel().getChunkSource().chunkMap;
+        io.papermc.paper.chunk.PlayerChunkLoader.PlayerLoaderData data = chunkMap.playerChunkManager.getData(this.getHandle());
+        if (data != null) {
+            if (data.nextChunkSendTarget != 0) {
+                long scheduleDelay = data.loader.getTargetSendPerPlayerAddend(this.getHandle());
+                if (scheduleDelay != 0) {
+                    long newTimeIfCurrentlyScheduled = System.nanoTime() + scheduleDelay;
+                    data.nextChunkSendTarget = Math.min(data.nextChunkSendTarget, newTimeIfCurrentlyScheduled);
+                }
+            }
+        }
+        // Martijn end - update chunk send deadline based on chunk send rate
     }
     // Martijn in Paper end - per-player chunk send rate
     // Paper end - implement view distances
