From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Sep 2021 15:38:51 +0200
Subject: [PATCH] Support lower chunk send rates


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index 3d0bdad12278d14c5f8084f3bc874ea2aaf99618..5047b010fc35fff6fa1c95e4929b239502f41a92 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -354,7 +354,7 @@ public final class PlayerChunkLoader {
     // Martijn in Paper start - per-player chunk send rate
     protected long getTargetSendPerPlayerAddend(ServerPlayer player) {
         final double chunkSendRate = player.getBukkitEntity().getTargetChunkSendRate();
-        // disable the limit for any given negative rate (with -1 the intended value to denote this)
+        // disable the limit for any given negative rate (with -1 the intended value to denote this, and 0L the appropriate output of this method in that case)
         return chunkSendRate < 0 ? 0L : (long) Math.round(1.0e9 / chunkSendRate);
     }
     // Martijn in Paper end - per-player chunk send rate
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0990142c93d3a60cd736b8fc80e6f5a17e35ae6d..9139e8bbe65423c78c517f7c6535b00726f42f95 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -632,7 +632,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         final Double playerTargetChunkSendRate = this.getHandle().targetChunkSendRate;
         final double effectiveTargetChunkSendRate = playerTargetChunkSendRate == null ? PaperConfig.playerTargetChunkSendRate : playerTargetChunkSendRate;
         // Make sure the returned rate is -1, or >= 1
-        return effectiveTargetChunkSendRate < 0 ? -1 : (effectiveTargetChunkSendRate <= 1 ? 1 : effectiveTargetChunkSendRate);
+        return effectiveTargetChunkSendRate < 0 ? -1 : Math.max(effectiveTargetChunkSendRate, 1.0e-6);
     }
 
     @Override
