From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Sep 2021 23:38:25 +0200
Subject: [PATCH] Update chunk send deadline based on chunk send rate


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index 3d0bdad12278d14c5f8084f3bc874ea2aaf99618..9775d05549bfc6541026dc11e198153b08605149 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -352,7 +352,7 @@ public final class PlayerChunkLoader {
     }
 
     // Martijn in Paper start - per-player chunk send rate
-    protected long getTargetSendPerPlayerAddend(ServerPlayer player) {
+    public long getTargetSendPerPlayerAddend(ServerPlayer player) { // Martijn - update chunk send deadline based on chunk send rate
         final double chunkSendRate = player.getBukkitEntity().getTargetChunkSendRate();
         // disable the limit for any given negative rate (with -1 the intended value to denote this)
         return chunkSendRate < 0 ? 0L : (long) Math.round(1.0e9 / chunkSendRate);
@@ -728,7 +728,7 @@ public final class PlayerChunkLoader {
         protected boolean usingLookingPriority;
 
         protected final ServerPlayer player;
-        protected final PlayerChunkLoader loader;
+        public final PlayerChunkLoader loader; // Martijn in Paper - update chunk send deadline based on chunk send rate
 
         // warning: modifications of this field must be aware that the loadQueue inside PlayerChunkLoader uses this field
         // in a comparator!
@@ -760,7 +760,7 @@ public final class PlayerChunkLoader {
         protected Boolean allowEntityTrackingBeyondNoTickViewDistance = null;
         // Martijn in Paper end - entity tracking view distance
 
-        protected long nextChunkSendTarget;
+        public long nextChunkSendTarget; // Martijn in Paper - update chunk send deadline based on chunk send rate
 
         public PlayerLoaderData(final ServerPlayer player, final PlayerChunkLoader loader) {
             this.player = player;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0990142c93d3a60cd736b8fc80e6f5a17e35ae6d..449fea3b6bb3c4ee91bb5187175066ea023f8746 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -638,6 +638,19 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     @Override
     public void setTargetChunkSendRate(Double chunkSendRate) {
         this.getHandle().targetChunkSendRate = chunkSendRate;
+        // Martijn start - update chunk send deadline based on chunk send rate
+        net.minecraft.server.level.ChunkMap chunkMap = this.getHandle().getLevel().getChunkSource().chunkMap;
+        io.papermc.paper.chunk.PlayerChunkLoader.PlayerLoaderData data = chunkMap.playerChunkManager.getData(this.getHandle());
+        if (data != null) {
+            if (data.nextChunkSendTarget != 0) {
+                long scheduleDelay = data.loader.getTargetSendPerPlayerAddend(this.getHandle());
+                if (scheduleDelay != 0) {
+                    long newTimeIfCurrentlyScheduled = System.nanoTime() + scheduleDelay;
+                    data.nextChunkSendTarget = Math.min(data.nextChunkSendTarget, newTimeIfCurrentlyScheduled);
+                }
+            }
+        }
+        // Martijn end - update chunk send deadline based on chunk send rate
     }
     // Martijn in Paper end - per-player chunk send rate
     // Paper end - implement view distances
