From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sat, 15 Jan 2022 19:26:03 +0100
Subject: [PATCH] Fix do not remove ticket manager if moving within a world


diff --git a/src/main/java/net/minecraft/server/level/DistanceManager.java b/src/main/java/net/minecraft/server/level/DistanceManager.java
index 2c8ce88f31808c39c4c60745220229e20184d043..31cf9d3241aca828afef183cab6edabc8b71741f 100644
--- a/src/main/java/net/minecraft/server/level/DistanceManager.java
+++ b/src/main/java/net/minecraft/server/level/DistanceManager.java
@@ -451,7 +451,6 @@ public abstract class DistanceManager {
     }
 
     public void removePlayer(SectionPos pos, ServerPlayer player, boolean keepTicketManager) { // Martijn
-        keepTicketManager = false; // Martijn temp
         ChunkPos chunkcoordintpair = pos.chunk();
         long i = chunkcoordintpair.toLong();
         ObjectSet<ServerPlayer> objectset = (ObjectSet) this.playersPerChunk.get(i);
