From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 23 Jan 2022 20:13:43 +0100
Subject: [PATCH] Make barrier and light blocks breakable in survival mode


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 334384654e87184fa7d29ab31b7408db56da0a11..7dd63bd7e8583118675542d84af20a50a3206141 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -235,7 +235,37 @@ public class ServerPlayerGameMode {
                 }
                 // CraftBukkit end
 
-                if (this.isCreative() || (player.getBukkitEntity().getGameMode() == GameMode.SURVIVAL && (player.getBukkitEntity().hasPermission("minecraft.debugstick.survivallike") || player.getBukkitEntity().hasPermission("minecraft.debugstick.always")) && player.getItemInHand(InteractionHand.MAIN_HAND).is(Items.DEBUG_STICK))) { // Martijn - survival debug stick
+                // Martijn start - survival debug stick, survival barrier blocks, survival light blocks
+                boolean canCreativeDestroy = false;
+                // Check for creative
+                if (this.isCreative()) {
+                    canCreativeDestroy = true;
+                } else {
+                    if (this.isSurvival()) {
+                        // Check for survival debug stick
+                        boolean isUsingSurvivalDebugStick = false;
+                        if (player.getBukkitEntity().hasPermission("minecraft.debugstick.survivallike") || player.getBukkitEntity().hasPermission("minecraft.debugstick.always")) {
+                            if (player.getItemInHand(InteractionHand.MAIN_HAND).is(Items.DEBUG_STICK)) {
+                                isUsingSurvivalDebugStick = true;
+                            }
+                        }
+                        if (isUsingSurvivalDebugStick) {
+                            canCreativeDestroy = true;
+                        } else {
+                            // Check for breaking barrier
+                            if (player.getItemInHand(InteractionHand.MAIN_HAND).is(Items.BARRIER) && this.level.getBlockState(pos).is(Blocks.BARRIER)) {
+                                canCreativeDestroy = true;
+                            } else {
+                                // Check for breaking light
+                                if (player.getItemInHand(InteractionHand.MAIN_HAND).is(Items.LIGHT) && this.level.getBlockState(pos).is(Blocks.LIGHT)) {
+                                    canCreativeDestroy = true;
+                                }
+                            }
+                        }
+                    }
+                }
+                if (canCreativeDestroy) {
+                    // Martijn end - survival debug stick, survival barrier blocks, survival light blocks
                     this.destroyAndAck(pos, action, "creative destroy");
                     return;
                 }
