From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 17 Jan 2022 12:37:38 +0100
Subject: [PATCH] Add utility methods for getting the distance between a player
 and a chunk


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index c032aec0e35f1be6ef967577f47dbc4a304515b7..09a7cb5632ad2cb0b9dde94f6d7ed7d4eb9c4058 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -392,6 +392,36 @@ public class ServerPlayer extends Player {
     }
     // Martijn end - getting the player's chunk
 
+    // Martijn start - getting distance to chunks
+
+    public int getManhattanDistanceToChunk(ChunkPos chunk) {
+        return Math.max(Math.abs(this.getChunkX() - chunk.x), Math.abs(this.getChunkZ() - chunk.z));
+    }
+
+    public int getManhattanDistanceToChunk(long chunkKey) {
+        return this.getManhattanDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    public long getSquaredPythagoreanDistanceToChunk(ChunkPos chunk) {
+        long dx = this.getChunkX() - chunk.x;
+        long dz = this.getChunkZ() - chunk.z;
+        return dx * dx + dz * dz;
+    }
+
+    public long getSquaredPythagoreanDistanceToChunk(long chunkKey) {
+        return this.getSquaredPythagoreanDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    public double getPythagoreanDistanceToChunk(ChunkPos chunkPos) {
+        return Math.sqrt(this.getSquaredPythagoreanDistanceToChunk(chunkPos));
+    }
+
+    public double getPythagoreanDistanceToChunk(long chunkKey) {
+        return this.getPythagoreanDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    // Martijn end - getting distance to chunks
+
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
     // If this is an issue, PRs are welcome
     public final BlockPos getSpawnPoint(ServerLevel worldserver) {
