From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 19 Jan 2022 02:52:21 +0100
Subject: [PATCH] Fix usage of other trackers in combined player tracker


diff --git a/src/main/java/net/minecraft/server/level/DistanceManager.java b/src/main/java/net/minecraft/server/level/DistanceManager.java
index 9ee2441cb104d5a8fab62b6c94eca1d1d4631540..7d2289cb28f6b5937cbb746aa4c0e9cbfafb419d 100644
--- a/src/main/java/net/minecraft/server/level/DistanceManager.java
+++ b/src/main/java/net/minecraft/server/level/DistanceManager.java
@@ -774,9 +774,10 @@ public abstract class DistanceManager {
             boolean withinSomeViewDistance = withinViewDistance;
             for (PlayerTicketTracker otherTracker : DistanceManager.this.getPlayerTicketManagers()) {
                 if (otherTracker != tracker) {
-                    minDistance = Math.min(distance, otherTracker.getLevel(pos));
-                    oldWithinSomeViewDistance |= otherTracker.haveTicketFor(distance);
-                    withinSomeViewDistance |= otherTracker.haveTicketFor(distance);
+                    int otherTrackerDistance = otherTracker.getLevel(pos);
+                    minDistance = Math.min(minDistance, otherTracker.getLevel(pos));
+                    oldWithinSomeViewDistance |= otherTracker.haveTicketFor(otherTrackerDistance);
+                    withinSomeViewDistance |= otherTracker.haveTicketFor(otherTrackerDistance);
                 }
             }
             this.onCombinedPlayerTicketTrackerLevelChange(pos, minDistance, oldWithinSomeViewDistance, withinSomeViewDistance);
