From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 9 Jun 2022 01:59:25 +0200
Subject: [PATCH] Wither skulls can give the wither effect on easy difficulty


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 5705af5bb77ae12c84f8129783687f6c63ac7e05..3f999625176b455474d1bcc89911113855abae0b 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -99,6 +99,15 @@ public class PaperWorldConfig {
 
     // Martijn end - cave spiders have poison on easy difficulty
 
+    // Martijn start - wither effect from wither skulls on easy difficulty
+
+    public boolean witherEffectFromWitherSkullsOnEasyDifficulty = false;
+    private void witherEffectFromWitherSkullsOnEasyDifficulty() {
+        witherEffectFromWitherSkullsOnEasyDifficulty = getBoolean("wither-effect-from-wither-skulls-on-easy-difficulty", witherEffectFromWitherSkullsOnEasyDifficulty);
+    }
+
+    // Martijn end - wither effect from wither skulls on easy difficulty
+
     public boolean zombiesTargetTurtleEggs = true;
     private void zombiesTargetTurtleEggs() {
         zombiesTargetTurtleEggs = getBoolean("zombies-target-turtle-eggs", zombiesTargetTurtleEggs);
diff --git a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
index 2867e841e73a3edfdeb83af9d96e0d0cd4116a68..261464c3aa542b0aef7eafdda97a38ae899ce873 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -75,7 +75,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
             if (flag && entity instanceof LivingEntity) {
                 byte b0 = 0;
 
-                if (this.level.getDifficulty() == Difficulty.NORMAL) {
+                if (this.level.getDifficulty() == Difficulty.NORMAL || (this.level.getDifficulty() == Difficulty.EASY && this.level.paperConfig.witherEffectFromWitherSkullsOnEasyDifficulty)) { // Martijn - wither effect from wither skulls on easy difficulty
                     b0 = 10;
                 } else if (this.level.getDifficulty() == Difficulty.HARD) {
                     b0 = 40;
