From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 8 Jun 2022 22:45:37 +0200
Subject: [PATCH] Configuration to remove invalid feature resource location
 console logs


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index f2660f4d5f798832fc44941929d1f9d1b5ce2b91..16b8db81e21037e85f275272c5191bcfd9e3d428 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -91,6 +91,15 @@ public class PaperConfig {
         }
     }
 
+    // Martijn start - remove invalid feature resource location console logs
+
+    public static boolean logInvalidFeatureResourceLocations = true;
+    private static void logInvalidFeatureResourceLocations() {
+        logInvalidFeatureResourceLocations = getBoolean("settings.log-invalid-feature-resource-locations", logInvalidFeatureResourceLocations);
+    }
+
+    // Martijn end - remove invalid feature resource location console logs
+
     // Martijn start - remove unrecognized recipe console logs
 
     public static boolean logUnrecognizedRecipes = true;
diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index fe6cf3787b9e4240d95f905fbc5969aded7652d0..d5bd021423fc06ad30b77659854ce58e0c9a9497 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -1092,7 +1092,11 @@ public class PaperWorldConfig {
 
             final net.minecraft.resources.ResourceLocation location = new net.minecraft.resources.ResourceLocation(key);
             if (!registry.containsKey(location)) {
-                logError("Invalid feature resource location: " + location);
+                // Martijn start - remove invalid feature resource location console logs
+                if (PaperConfig.logInvalidFeatureResourceLocations) {
+                    logError("Invalid feature resource location: " + location);
+                }
+                // Martijn end - remove invalid feature resource location console logs
                 continue;
             }
 
