From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 27 Jun 2022 22:58:25 +0200
Subject: [PATCH] Z


diff --git a/src/main/java/nl/martijnmuijsers/paper/antixray/ChunkPacketBlockControllerCustom.java b/src/main/java/nl/martijnmuijsers/paper/antixray/ChunkPacketBlockControllerCustom.java
index a9c26bcff171ddb6aaafa26b2e6057c84ae65d60..f61d1988fe97941aa53fc62a58776af21e0a3773 100644
--- a/src/main/java/nl/martijnmuijsers/paper/antixray/ChunkPacketBlockControllerCustom.java
+++ b/src/main/java/nl/martijnmuijsers/paper/antixray/ChunkPacketBlockControllerCustom.java
@@ -75,7 +75,8 @@ public final class ChunkPacketBlockControllerCustom extends ChunkPacketBlockCont
         bitStorageWriter.setBuffer(chunkPacketInfoAntiXray.getBuffer());
 
         for (int chunkSectionIndex = 0; chunkSectionIndex < sectionsCount; chunkSectionIndex++) {
-            if (chunkPacketInfoAntiXray.isWritten(chunkSectionIndex) && chunkPacketInfoAntiXray.getPresetValues(chunkSectionIndex) != null) {
+            MinecraftServer.LOGGER.info(chunkPacketInfoAntiXray.isWritten(chunkSectionIndex) + " --- " + (chunkPacketInfoAntiXray.getPresetValues(chunkSectionIndex) != null));
+            if (chunkPacketInfoAntiXray.isWritten(chunkSectionIndex)) {
 
                 bitStorageReader.setBits(chunkPacketInfoAntiXray.getBits(chunkSectionIndex));
                 bitStorageReader.setIndex(chunkPacketInfoAntiXray.getIndex(chunkSectionIndex));
