From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 3 Jul 2022 00:16:28 +0200
Subject: [PATCH] Add Dirty Stone Bricks


diff --git a/src/main/java/nl/martijnmuijsers/paper/block/SuCraftBlocks.java b/src/main/java/nl/martijnmuijsers/paper/block/SuCraftBlocks.java
index 7f9f82567336ce22466d1226884a06dfa5782224..88c4c7cc606c6e65714d57e4af634504389443b2 100644
--- a/src/main/java/nl/martijnmuijsers/paper/block/SuCraftBlocks.java
+++ b/src/main/java/nl/martijnmuijsers/paper/block/SuCraftBlocks.java
@@ -40,6 +40,24 @@ public class SuCraftBlocks {
         SuCraftBlockLoot::dropSelf
     );
 
+    /**
+     * Based on {@link #SNOWY_STONE_BRICKS}
+     */
+    public static final Block DIRTY_STONE_BRICKS = register(
+        "dirty_stone_bricks",
+        BlockReplacementRule.stateless()
+            .withHost(Blocks.INFESTED_CRACKED_STONE_BRICKS)
+            .withHostReplacement(Blocks.CRACKED_STONE_BRICKS)
+            .withFullBlockBoundingBoxHostItem()
+            .withFallback(Blocks.STONE_BRICKS)
+            .withFallbackBlockAsFallbackItem()
+            .withHostReplacingItemFromHostBlock()
+            .withHostReplacingItemReplacementFromHostReplacementBlock()
+            .withModel(Model.DEFAULT_CUBE_MODEL(ResourcePackAssetIntention.BLOCK)),
+        new Block(BlockBehaviour.Properties.of(Material.STONE).requiresCorrectToolForDrops().strength(1.5F, 6.0F)),
+        SuCraftBlockLoot::dropSelf
+    );
+
     private static @NotNull Block register(@NotNull String id, @NotNull BlockReplacementRule.Builder<?> replacementRuleBuilder, @NotNull Block block, @Nullable Supplier<@Nullable LootTable> lootTableSupplier) {
         return register(id, replacementRuleBuilder, block, lootTableSupplier == null ? null : ($ -> lootTableSupplier.get()));
     }
diff --git a/src/main/java/nl/martijnmuijsers/paper/item/SuCraftItems.java b/src/main/java/nl/martijnmuijsers/paper/item/SuCraftItems.java
index 50501a6736d9d2d097c31789f97cf318eee91f7c..50d5036c273b87a0b3d19a777c97f8a3591c1c17 100644
--- a/src/main/java/nl/martijnmuijsers/paper/item/SuCraftItems.java
+++ b/src/main/java/nl/martijnmuijsers/paper/item/SuCraftItems.java
@@ -29,6 +29,11 @@ public class SuCraftItems {
      */
     public static final Item SNOWY_STONE_BRICKS = registerBlock(SuCraftBlocks.SNOWY_STONE_BRICKS, "Snowy Stone Bricks", CreativeModeTab.TAB_BUILDING_BLOCKS);
 
+    /**
+     * Based on {@link Items#STONE_BRICKS}
+     */
+    public static final Item DIRTY_STONE_BRICKS = registerBlock(SuCraftBlocks.DIRTY_STONE_BRICKS, "Dirty Stone Bricks", CreativeModeTab.TAB_BUILDING_BLOCKS);
+
     public static Item registerBlock(Block block, String displayName, CreativeModeTab group) {
         Item registeredItem = Items.registerBlock(block, group);
         block.customItemForBlock = registeredItem;
