From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Fri, 5 Aug 2022 14:39:53 +0200
Subject: [PATCH] No signed chat


diff --git a/src/main/java/io/papermc/paper/adventure/ChatProcessor.java b/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
index 46d1ed7e617abc7096f5d7a391b138238f42aaa4..7541e53c83be8609a90d0092777abb821c5b37bc 100644
--- a/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
+++ b/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
@@ -310,7 +310,7 @@ public final class ChatProcessor {
 
         private void sendToServer(final ChatType.Bound chatType, final @Nullable Function<Audience, net.minecraft.network.chat.Component> msgFunction) {
             final PlayerChatMessage toConsoleMessage = msgFunction == null ? ChatProcessor.this.message : ChatProcessor.this.message.withUnsignedContent(msgFunction.apply(ChatProcessor.this.server.console));
-            ChatProcessor.this.server.logChatMessage(toConsoleMessage.serverContent(), chatType, ChatProcessor.this.server.getPlayerList().verifyChatTrusted(toConsoleMessage, ChatProcessor.this.player.asChatSender()) ? null : "Not Secure");
+            ChatProcessor.this.server.logChatMessage(toConsoleMessage.serverContent(), chatType, null); // Martijn - no signed chat
         }
 
         record ConstantFunction(net.minecraft.network.chat.Component component) implements Function<Audience, net.minecraft.network.chat.Component> {
diff --git a/src/main/java/net/minecraft/network/chat/OutgoingPlayerChatMessage.java b/src/main/java/net/minecraft/network/chat/OutgoingPlayerChatMessage.java
index fc43cc6f5bbb3485b9ddcc441a13399e57289586..cd3a54e160025da8981e692eabdc8c9304c3b508 100644
--- a/src/main/java/net/minecraft/network/chat/OutgoingPlayerChatMessage.java
+++ b/src/main/java/net/minecraft/network/chat/OutgoingPlayerChatMessage.java
@@ -51,7 +51,7 @@ public interface OutgoingPlayerChatMessage {
             if (!playerChatMessage.isFullyFiltered()) {
                 RegistryAccess registryAccess = serverPlayer.level.registryAccess();
                 ChatType.BoundNetwork boundNetwork = bound.toNetwork(registryAccess);
-                serverPlayer.connection.send(new ClientboundPlayerChatPacket(playerChatMessage, boundNetwork));
+                serverPlayer.connection.send(ClientboundPlayerChatPacket.createUnsigned(serverPlayer, playerChatMessage, boundNetwork)); // Martijn - no signed chat
                 serverPlayer.connection.addPendingMessage(playerChatMessage);
             }
 
@@ -90,7 +90,7 @@ public interface OutgoingPlayerChatMessage {
                 this.playersWithFullMessage.add(serverPlayer);
                 RegistryAccess registryAccess = serverPlayer.level.registryAccess();
                 ChatType.BoundNetwork boundNetwork = bound.toNetwork(registryAccess);
-                serverPlayer.connection.send(new ClientboundPlayerChatPacket(playerChatMessage, boundNetwork), PacketSendListener.exceptionallySend(() -> {
+                serverPlayer.connection.send(ClientboundPlayerChatPacket.createUnsigned(serverPlayer, playerChatMessage, boundNetwork), PacketSendListener.exceptionallySend(() -> { // Martijn - no signed chat
                     return new ClientboundPlayerChatHeaderPacket(this.message);
                 }));
                 serverPlayer.connection.addPendingMessage(playerChatMessage);
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerChatPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerChatPacket.java
index 0b818969d74eae05e2d2adfdbc38e513972c61a1..80e1bbe1c8da3271aa1acfd51203f1f9956c6741 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerChatPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerChatPacket.java
@@ -4,8 +4,11 @@ import java.util.Optional;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.ChatType;
+import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.PlayerChatMessage;
 import net.minecraft.network.protocol.Packet;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
 
 public record ClientboundPlayerChatPacket(PlayerChatMessage message, ChatType.BoundNetwork chatType) implements Packet<ClientGamePacketListener> {
     public ClientboundPlayerChatPacket(FriendlyByteBuf buf) {
@@ -31,4 +34,22 @@ public record ClientboundPlayerChatPacket(PlayerChatMessage message, ChatType.Bo
     public Optional<ChatType.Bound> resolveChatType(RegistryAccess dynamicRegistryManager) {
         return this.chatType.resolve(dynamicRegistryManager);
     }
+
+    // Martijn start - no signed chat
+
+    public static Packet<ClientGamePacketListener> createUnsigned(ServerPlayer player, PlayerChatMessage message, ChatType.BoundNetwork chatType) {
+        // Based on FreedomChat: ru.bk.oharass.freedomchat.FreedomChat.on(PlayerJoinEvent)
+        final Component content = message.unsignedContent().orElse(message.signedContent().decorated());
+
+        final Optional<ChatType.Bound> ctbo = chatType.resolve(player.level.registryAccess());
+        if (ctbo.isEmpty()) {
+            MinecraftServer.LOGGER.warn("Processing packet with unknown ChatType " + chatType.chatType(), new Throwable());
+            return new ClientboundPlayerChatPacket(message, chatType);
+        }
+        final Component decoratedContent = ctbo.orElseThrow().decorate(content);
+        return new ClientboundSystemChatPacket(decoratedContent, false);
+    }
+
+    // Martijn end - no signed chat
+
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 98d9900aecaf974945858f22a9647bf34571a25a..7d144a970b2f9f718d6b6e5aa087ae4567a9815f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2306,10 +2306,11 @@ public class ServerPlayer extends Player {
     }
 
     public void sendChatHeader(SignedMessageHeader header, MessageSignature headerSignature, byte[] bodyDigest) {
-        if (this.acceptsChatMessages()) {
-            this.connection.send(new ClientboundPlayerChatHeaderPacket(header, headerSignature, bodyDigest));
-        }
-
+        // Martijn start - no signed chat
+//        if (this.acceptsChatMessages()) {
+//            this.connection.send(new ClientboundPlayerChatHeaderPacket(header, headerSignature, bodyDigest));
+//        }
+        // Martijn end - no signed chat
     }
 
     public String getIpAddress() {
@@ -2375,7 +2376,7 @@ public class ServerPlayer extends Player {
     }
 
     public void sendServerStatus(ServerStatus metadata) {
-        this.connection.send(new ClientboundServerDataPacket(metadata.getDescription(), metadata.getFavicon(), metadata.previewsChat(), metadata.enforcesSecureChat()));
+        this.connection.send(new ClientboundServerDataPacket(metadata.getDescription(), metadata.getFavicon(), false/*metadata.previewsChat()*/, true/*metadata.enforcesSecureChat()*/)); // Martijn - no signed chat
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 99a18eb1aebfd52cfc51596ab3d82abf65aeb301..7b183bf877616903d5d8361ca679ce98fb49686d 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1433,7 +1433,7 @@ public abstract class PlayerList {
         // Paper end
         boolean flag = this.verifyChatTrusted(playerchatmessage, chatsender);
 
-        this.server.logChatMessage((unsignedFunction == null ? playerchatmessage : playerchatmessage.withUnsignedContent(unsignedFunction.apply(this.server.console))).serverContent(), chatmessagetype_a, flag ? null : "Not Secure"); // Paper
+        this.server.logChatMessage((unsignedFunction == null ? playerchatmessage : playerchatmessage.withUnsignedContent(unsignedFunction.apply(this.server.console))).serverContent(), chatmessagetype_a, null); // Paper // Martijn - no signed chat
         OutgoingPlayerChatMessage outgoingplayerchatmessage = OutgoingPlayerChatMessage.create(playerchatmessage);
         boolean flag1 = playerchatmessage.isFullyFiltered();
         boolean flag2 = false;
