From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Fri, 5 Aug 2022 12:04:44 +0200
Subject: [PATCH] Structure API / asdfuasf


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 03d541ff73aac3390d4c44266aafef935a4f7f1e..8911a868b97057bc8587ff7a6d528526bc389bc2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -62,7 +62,6 @@ import net.minecraft.world.level.StructureManager;
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.level.chunk.ImposterProtoChunk;
-import net.minecraft.world.level.levelgen.structure.Structure;
 import net.minecraft.world.level.levelgen.structure.StructureStart;
 import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.phys.AABB;
@@ -245,7 +244,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         ChunkPos chunkPos = ((CraftChunk) block.getChunk()).getHandle().getPos();
         StructureManager featureManager = this.getHandle().structureManager();
 
-        final Set<Structure> features;
+        final Set<net.minecraft.world.level.levelgen.structure.Structure> features;
         if (structureType != null) {
 //            // Base on CraftWorld.locateNearestStructure
 //            features = this.structureCache.computeIfAbsent(structureType, (type) -> {
