From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 24 Aug 2022 17:36:50 +0200
Subject: [PATCH] Ignore empty chat messages


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d106003d7286937ea08f4d467f64e49066860d34..cb6729c9c4f083e083996dc9d420ef72168973a0 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2596,7 +2596,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         // CraftBukkit start
         String s = message.signedContent().plain();
         if (s.isEmpty()) {
-            ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message");
+//            ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message"); // Martijn - ignore empty chat messages
         } else if (this.getCraftPlayer().isConversing()) {
             final String conversationInput = s;
             this.server.processQueue.add(new Runnable() {
