From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Fri, 5 Aug 2022 13:27:12 +0200
Subject: [PATCH] Custom blocks and items / asdjioasdoaaa


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java
index 903fb16d7851501ae1f25d9d3e84e0b2ff606e5b..0e4b059d57d021eed38df098395cebb4b1e8b97e 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundForgetLevelChunkPacket.java
@@ -58,7 +58,7 @@ public class ClientboundForgetLevelChunkPacket implements Packet<ClientGamePacke
                     callback.onSuccess();
                 }
                 if (connection.isConnected()) {
-                    ChunkPos chunkPos = new ChunkPos(ClientboundBlockUpdatePacket.this.x, ClientboundBlockUpdatePacket.this.z);
+                    ChunkPos chunkPos = new ChunkPos(ClientboundForgetLevelChunkPacket.this.x, ClientboundForgetLevelChunkPacket.this.z);
                     for (int y = connection.getPlayer().getLevel().getMinSection(); y <= connection.getPlayer().getLevel().getMaxSection(); y++) {
                         connection.despawnBlockEntitiesInChunkSection(SectionPos.of(chunkPos, y).asLong());
                     }
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
index 86dde845b1b68a046cbb5f9422df8909711ecf3a..ff5fe2b10be8105776bc270f910dae3c77292a63 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -115,11 +115,11 @@ public class ClientboundLevelChunkWithLightPacket implements Packet<ClientGamePa
                 }
                 if (connection.isConnected()) {
                     try {
-                        @Nullable LevelChunk chunk = connection.getPlayer().getLevel().getChunkIfLoaded(ClientboundBlockUpdatePacket.this.x, ClientboundBlockUpdatePacket.this.z);
+                        @Nullable LevelChunk chunk = connection.getPlayer().getLevel().getChunkIfLoaded(ClientboundLevelChunkWithLightPacket.this.x, ClientboundLevelChunkWithLightPacket.this.z);
                         if (chunk != null) {
                             int sectionsCount = chunk.getSectionsCount();
                             for (int sectionYIndex = 0; sectionYIndex < sectionsCount; sectionYIndex++) {
-                                SectionPos sectionPos = SectionPos.of(ClientboundBlockUpdatePacket.this.x, chunk.getSectionYFromSectionIndex(sectionYIndex), ClientboundBlockUpdatePacket.this.z);
+                                SectionPos sectionPos = SectionPos.of(ClientboundLevelChunkWithLightPacket.this.x, chunk.getSectionYFromSectionIndex(sectionYIndex), ClientboundLevelChunkWithLightPacket.this.z);
                                 connection.updatePotentialBlockEntitiesInSection(sectionPos, false, chunk);
                             }
                         }
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
index f5a02991c725d0b4b908145888825eda6bb4001a..835954827bd9f57685ccb06e29c9d5b7fad50d52 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -164,7 +164,7 @@ public class ClientboundSectionBlocksUpdatePacket implements Packet<ClientGamePa
                     callback.onSuccess();
                 }
                 if (connection.isConnected()) {
-                    connection.updatePotentialBlockEntitiesInSection(ClientboundBlockUpdatePacket.this.sectionPos, true, null);
+                    connection.updatePotentialBlockEntitiesInSection(ClientboundSectionBlocksUpdatePacket.this.sectionPos, true, null);
                 }
 
             }
